opendevin/memory/history.py

```diff
@@ -1,54 +1,260 @@
         """
-        if not isinstance(event_dict, dict):
-            raise AgentEventTypeError()
-        self.events.append(event_dict)
+        Return the last action from the event stream, filtered to exclude unwanted events.
+        """
+        # from end_id in reverse, find the first action
+        end_id = self._event_stream.get_latest_event_id() if end_id == -1 else end_id
+
+        last_action = next(
+            (
+                event
+                for event in self._event_stream.get_events(
+                    end_id=end_id, reverse=True, filter_out_type=self.filter_out
+                )
+                if isinstance(event, Action)
+            ),
+            None,
+        )
 
-    def get_events(self):
+        return last_action
+
+    def get_last_observation(self, end_id: int = -1) -> Observation | None:
+        """
+        Return the last observation from the event stream, filtered to exclude unwanted events.
         """
-        Get the events in the agent's recent history.
+        # from end_id in reverse, find the first observation
+        end_id = self._event_stream.get_latest_event_id() if end_id == -1 else end_id
+
+        last_observation = next(
+            (
+                event
+                for event in self._event_stream.get_events(
+                    end_id=end_id, reverse=True, filter_out_type=self.filter_out
+                )
+                if isinstance(event, Observation)
+            ),
+            None,
+        )
 
-        Returns:
-        - List: The list of events that the agent remembers easily.
+        return last_observation
+
+    def get_latest_user_message(self) -> str:
+        """
+        Return the latest user message from the event stream.
         """
-        return self.events
 
-    def get_total_length(self):
+        last_user_message = next(
+            (
+                event.content
+                for event in self._event_stream.get_events(reverse=True)
+                if isinstance(event, MessageAction) and event.source == EventSource.USER
+            ),
+            None,
+        )
+
+        return last_user_message if last_user_message is not None else ''
+
+    def get_last_events(self, n: int) -> list[Event]:
         """
-        Gives the total number of characters in all history
+        Return the last n events from the event stream.
+        """
+        # dummy agent is using this
+        # it should work, but it's not great to store temporary lists now just for a test
+        end_id = self._event_stream.get_latest_event_id()
+        start_id = max(0, end_id - n + 1)
+
+        return list(
+            event
+            for event in self._event_stream.get_events(
+                start_id=start_id,
+                end_id=end_id,
+                filter_out_type=self.filter_out,
+            )
+        )
 
-        Returns:
-        - Int: Total number of characters of the recent history.
+    def get_latest_event_id(self):
```

Comment:
"last" and "latest", potato, potatoe? 🤣

Reply:
This method has become an orphan it seems, it isn't actually used. Removed!